firedrake/mesh.py

```diff
@@ -617,20 +621,47 @@ def cell_set(self):
         return op2.Set(size, "%s_cells" % self.name, comm=self.comm)
 
-    def cell_subset(self, subdomain_id):
+    def cell_subset(self, subdomain_id, all_integer_subdomain_ids=()):
         """Return a subset over cells with the given subdomain_id."""
-        # FIXME: This is WRONG in the case of 1*(dx + dx(1))
-        # "otherwise" is "everywhere" - set(explicit subdomain ids)
-        if subdomain_id in ["everywhere", "otherwise"]:
+        if subdomain_id == "everywhere":
             return self.cell_set
+        if subdomain_id == "otherwise":
+            if len(all_integer_subdomain_ids) == 0:
+                return self.cell_set
+            key = ("otherwise", ) + all_integer_subdomain_ids
+        else:
+            key = subdomain_id
         try:
-            return self._subsets[subdomain_id]
+            return self._subsets[key]
         except KeyError:
-            indices = dmplex.get_cell_markers(self._plex,
-                                              self._cell_numbering,
-                                              subdomain_id)
-            self._subsets[subdomain_id] = op2.Subset(self.cell_set, indices)
-            return self._subsets[subdomain_id]
+            if subdomain_id == "otherwise":
+                ids = tuple(dmplex.get_cell_markers(self._plex,
+                                                    self._cell_numbering,
+                                                    sid)
+                            for sid in all_integer_subdomain_ids)
+                to_remove = np.unique(np.concatenate(ids))
+                indices = np.arange(self.cell_set.total_size, dtype=np.int32)
+                indices = np.delete(indices, to_remove)
+            else:
+                indices = dmplex.get_cell_markers(self._plex,
+                                                  self._cell_numbering,
+                                                  subdomain_id)
+            return self._subsets.setdefault(key, op2.Subset(self.cell_set, indices))
+
+    def measure_set(self, integral_type, subdomain_id,
+                    all_integer_subdomain_ids=()):
+        if integral_type == "cell":
+            return self.cell_subset(subdomain_id, all_integer_subdomain_ids)
+        elif integral_type in ("exterior_facet", "exterior_facet_vert",
+                               "exterior_facet_top", "exterior_facet_bottom"):
+            return self.exterior_facets.measure_set(integral_type, subdomain_id,
+                                                    all_integer_subdomain_ids)
+        elif integral_type in ("interior_facet", "interior_facet_vert",
+                               "interior_facet_horiz"):
+            return self.interior_facets.measure_set(integral_type, subdomain_id,
+                                                    all_integer_subdomain_ids)
+        else:
+            raise ValueError("Unknown integral type '%s'" % integral_type)
```

Comment:
`MeshTopology` learns about extruded measures.  Should those cases move to `ExtrudedMeshTopology`.